toqito/matrices/operator_sinkhorn.py

```diff
@@ -0,0 +1,198 @@
+
+    >>> print("rho_random", rho_random, sep='\n')
+    rho_random
+    [[0.23434155+0.j         0.20535572-0.04701708j 0.11523158-0.02017518j
+      0.18524981-0.13636277j]
+     [0.20535572+0.04701708j 0.2544268 +0.j         0.14478708-0.02004061j
+      0.22496295-0.11837381j]
+     [0.11523158+0.02017518j 0.14478708+0.02004061j 0.11824612+0.j
+      0.09088248-0.05551508j]
+     [0.18524981+0.13636277j 0.22496295+0.11837381j 0.09088248+0.05551508j
+      0.39298553+0.j        ]]
+
+    >>> sigma, F = operator_sinkhorn(rho=rho_random, dim=[2, 2])
+
+    This returns the result density matrix along with the operations list F.
+    SIGMA has all of its (single-party) reduced density matrices
+    proportional to the identity, and SIGMA = Tensor(F)*RHO*Tensor(F)'. In
+    other words, F contains invertible local operations that demonstrate
+    that RHO and SIGMA are locally equivalent.
+
+    >>> print("sigma", sigma, sep='\n')
+    sigma
+    [[ 0.34784186+0.j          0.09278034+0.00551919j -0.02275152+0.05104798j
+       0.20443565-0.12511156j]
+     [ 0.09278034-0.00551919j  0.15215814+0.j          0.1039554 -0.0350973j
+       0.02275152-0.05104798j]
+     [-0.02275152-0.05104798j  0.1039554 +0.0350973j   0.15215814+0.j
+      -0.09278034-0.00551919j]
+     [ 0.20443565+0.12511156j  0.02275152+0.05104798j -0.09278034+0.00551919j
+       0.34784186+0.j        ]]
+
+    >>> print("F", F, sep='\n')
+    F
+    [array([[ 1.33132101+0.00313432j, -0.38927801+0.14985847j],
+           [-0.37199021-0.14296461j,  1.24635929-0.00300407j]]), array([[ 1.16317427+0.00765735j, -0.16925696+0.09054309j],
+           [-0.18974418-0.09028502j,  0.93359381-0.00777829j]])]
+
+    Similarly to perform sinkhorn operation on multipartite state where the
+    subsystems are of dimensions 2, 3 and 4.
+
+    >>> # rho_random = random_density_matrix(24)
+    >>> # sigma, F = operator_sinkhorn(rho=rho_random, dim=[2, 3, 4])
+    >>> # print(sigma, F)
+
+    References
+    ==========
+    .. bibliography::
+        :filter: docname in docnames
+
+    :param rho: input density matrix of a multipartite system
+    :param dim: `None` list containing dimensions of each subsystem.
+                Assumes 2 subsystems with equal dimensions as default.
+    :param tol: `np.sqrt(np.finfo(float).eps)` Convergence tolerance of the iterative Sinkhorn Algorithm.
+                Assumes square root of numpy eps as default.
+    :raises: ValueError: if input density matrix is not a square matrix.
+    :raises: ValueError: if the product of dimensions provided/assumed does not match the dimension of density matrix.
+    :raises: RuntimeError: if the sinkhorn algorithm doesnot converge before 10000 iterations.
+    :raises: ValueError: if the density matrix provided is singular (or is not of full rank).
+
+    """ # noqa: E501
+    # Run checks on the input density matrix
+    rho = np.asarray(rho)
+    if rho.ndim != 2 or rho.shape[0] != rho.shape[1]:
```

Comment:
We already have a function that checks whether a matrix is square in `matrix_props.is_square`.